sharding/xorValidatorShuffler_test.go

```diff
@@ -0,0 +1,778 @@
+}
+
+func testDistributeValidators(
+	t *testing.T,
+	initialMap map[uint32][]Validator,
+	resultedMap map[uint32][]Validator,
+	distributedNodes []Validator,
+) {
+	totalResultingValidators := make([]Validator, 0)
+	totalLen := 0
+	for _, valList := range resultedMap {
+		totalResultingValidators = append(totalResultingValidators, valList...)
+		totalLen += len(valList)
+	}
+
+	totalValidators := make([]Validator, 0)
+	for _, valList := range initialMap {
+		totalValidators = append(totalValidators, valList...)
+	}
+	assert.Equal(t, len(totalValidators)+len(distributedNodes), totalLen)
+
+	totalValidators = append(totalValidators, distributedNodes...)
+	assert.True(t, contains(totalResultingValidators, totalValidators))
+}
+
+func numberMatchingNodes(searchList []Validator, toFind []Validator) int {
+	nbFound := 0
+	for _, v1 := range toFind {
+		for _, v2 := range searchList {
+			if v1 == v2 {
+				nbFound++
+				break
+			}
+		}
+	}
+
+	return nbFound
+}
+
+func testLeaving(
+	t *testing.T,
+	eligible map[uint32][]Validator,
+	waiting map[uint32][]Validator,
+	prevLeaving []Validator,
+	newLeaving []Validator,
+) (int, map[uint32]int) {
+	nbLeavingPerShard := make(map[uint32]int)
+
+	nbLeavingFromEligible := 0
+	for i, eligibleList := range eligible {
+		nbWantingToLeaveFromList := numberMatchingNodes(eligibleList, prevLeaving)
+		maxAllowedToLeaveFromList := len(waiting[i])
+		nbLeaving := nbWantingToLeaveFromList
+		if nbLeaving > maxAllowedToLeaveFromList {
+			nbLeaving = maxAllowedToLeaveFromList
+		}
+
+		nbLeavingPerShard[i] += nbLeaving
+		nbLeavingFromEligible += nbLeaving
+	}
+	assert.Equal(t, nbLeavingFromEligible, len(prevLeaving)-len(newLeaving))
+
+	return nbLeavingFromEligible, nbLeavingPerShard
+}
+
+func testShuffledOut(
+	t *testing.T,
+	eligibleMap map[uint32][]Validator,
+	waitingMap map[uint32][]Validator,
+	newEligible map[uint32][]Validator,
+	shuffledOut []Validator,
+	prevleaving []Validator,
+	newleaving []Validator,
+) {
+	nbAllLeaving, _ := testLeaving(t, eligibleMap, waitingMap, prevleaving, newleaving)
+	allWaiting := getValidatorsInMap(waitingMap)
+	allEligible := getValidatorsInMap(eligibleMap)
+	assert.Equal(t, len(shuffledOut)+nbAllLeaving, len(allWaiting))
+
+	allNewEligible := getValidatorsInMap(newEligible)
+	assert.Equal(t, len(allEligible)-len(shuffledOut)-nbAllLeaving, len(allNewEligible))
+
+	newNodes := append(allNewEligible, shuffledOut...)
+	assert.NotEqual(t, allEligible, newNodes)
+	assert.True(t, contains(newNodes, allEligible))
+}
+
+func createDefaultXorShuffler() *randXORShuffler {
+	return NewXorValidatorsShuffler(
+		100,
+		100,
+		0.2,
+		false,
+	)
+}
+
+func getValidatorsInMap(valMap map[uint32][]Validator) []Validator {
+	result := make([]Validator, 0)
+
+	for _, valList := range valMap {
+		result = append(result, valList...)
+	}
+
+	return result
+}
+
+func Test_xorBytes_SameLen(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes(firstArray, secondArray)
```

Comment:
maybe change rez to result, valid in the whole file.

Reply:
done